x/ccv/provider/keeper/partial_set_security.go

```diff
@@ -53,13 +51,89 @@ func (k Keeper) HandleOptOut(ctx sdk.Context, chainID string, providerAddr types
 			"opting out of an unknown or not running consumer chain, with id: %s", chainID)
 	}
 
-	if k.IsToBeOptedIn(ctx, chainID, providerAddr) {
-		// a validator to be opted out cancels out a validator to be opted in
-		k.DeleteToBeOptedIn(ctx, chainID, providerAddr)
-	} else if !k.IsToBeOptedOut(ctx, chainID, providerAddr) && k.IsOptedIn(ctx, chainID, providerAddr) {
-		// a validator can only be set for opt out if it is opted in and not already set for opt out
-		k.SetToBeOptedOut(ctx, chainID, providerAddr)
+	if topN, found := k.GetTopN(ctx, chainID); found {
+		// a validator cannot opt out from a Top N chain if the validator is in the Top N validators
+		validator, validatorFound := k.stakingKeeper.GetValidatorByConsAddr(ctx, providerAddr.ToSdkConsAddr())
+		if !validatorFound {
+			return errorsmod.Wrapf(
+				stakingtypes.ErrNoValidatorFound,
+				"validator with consensus address %s could not be found", providerAddr.ToSdkConsAddr())
+		}
+		power := k.stakingKeeper.GetLastValidatorPower(ctx, validator.GetOperator())
+		minPowerToOptIn := k.ComputeMinPowerToOptIn(ctx, k.stakingKeeper.GetLastValidators(ctx), topN)
+
+		if power >= minPowerToOptIn {
+			return errorsmod.Wrapf(
+				types.ErrCannotOptOutFromTopN,
+				"validator with power (%d) cannot opt out from Top N chain because all validators"+
+					"with at least %d power have to validate", power, minPowerToOptIn)
+		}
 	}
 
+	k.DeleteOptedIn(ctx, chainID, providerAddr)
 	return nil
 }
+
+// OptInTopNValidators opts in to `chainID` all the `bondedValidators` that have at least `minPowerToOptIn` power
+func (k Keeper) OptInTopNValidators(ctx sdk.Context, chainID string, bondedValidators []stakingtypes.Validator, minPowerToOptIn int64) {
+	for _, val := range bondedValidators {
+		power := k.stakingKeeper.GetLastValidatorPower(ctx, val.GetOperator())
+		if power >= minPowerToOptIn {
+			consAddr, err := val.GetConsAddr()
+			if err != nil {
+				k.Logger(ctx).Error("could not retrieve validators consensus address",
+					"validator", val,
+					"error", err)
+				continue
+			}
+
+			// if validator already exists it gets overwritten
+			k.SetOptedIn(ctx, chainID, types.NewProviderConsAddress(consAddr))
+		} // else validators that do not belong to the top N validators but were opted in, remain opted in
+	}
+}
+
+// ComputeMinPowerToOptIn returns the minimum power needed for a validator (from the bonded validators)
+// to belong to the `topN` validators
+func (k Keeper) ComputeMinPowerToOptIn(ctx sdk.Context, bondedValidators []stakingtypes.Validator, topN uint32) int64 {
```

Comment:
Hey,

I think there is a bug here.
If topN == 0, topNThreshold will be the maximal power.

However, this is invoked e.g. here https://github.com/cosmos/interchain-security/blob/30b99aa8eb193a08bef96d181fb21c13ddcc9ca5/x/ccv/provider/keeper/partial_set_security.go#L65 to check whether someone can opt out.
This disallows the val with maximal power from opting out, for example.

I think it should just return something to signify infinity if topN == 0

Reply:
Great catch. I fixed the condition on `HandleOptOut` to:
```
if topN, found := k.GetTopN(ctx, chainID); found && topN > 0 {
```
to guarantee that `ComputeMinPowerToOptIn` does not get `topN == 0`.
But, also changed `ComputeMinPowerToOptIn` to return `math.MaxInt64` when `topN == 0` to avoid potential pitfalls.